### PR TITLE
Introduce PySam as an alternative parsing library.

### PR DESCRIPTION
**Pull request by tneymanov**
Uses multi threading and two pipes - to send data from parent thread to child, and from child into PySam class VariantFile (which is used by parent to read records). This is required to comply and to not overwrite base methods in VcfParser.

In order to send a line at a time in each and parse them one by one, readline() method is used - alternative methods deadlock, as they expect EOF before they can proceed. Therefore, every line read should be accompanied with a break-line character.

Due to quota issues, haven't checked the correctness of the integration tests yet, but manual run works and produces a table. Will update once quota is available and integration tests are run.

### CODE REVIEW

**Comment by allieychen** [comment]
Have you tried running tests on a large dataset yet? How is the speed?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 1]
```diff
@@ -504,6 +508,162 @@ def _get_variant_calls(self, record, formats):
...
+    try:
+      self._to_child.write(data_line + '\n')
+      self._to_child.flush()
+      return self._convert_to_variant(next(self._vcf_reader))
+    except (MemoryError, IOError) as e:
+      logging.warning('VCF record read failed in %s for line %s: %s',
+                      self._file_name, data_line, str(e))
+      return MalformedVcfRecord(self._file_name, data_line, str(e))
+
+  def _convert_to_variant(
```
Comment (allieychen) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: I think one line is easier to read (put `self` and `record` in the same line).

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 2]
```diff
@@ -504,6 +508,162 @@ def _get_variant_calls(self, record, formats):
...
+    p_read, c_write = os.pipe()
+    c_read, p_write = os.pipe()
+
+    processid = os.fork()
+    if processid:
+      os.close(c_read)
+      os.close(c_write)
+      into_pysam = os.fdopen(p_read)
+      self._to_child = os.fdopen(p_write, 'w')
+      self._vcf_reader = libcbcf.VariantFile(into_pysam, 'r')
```
Comment (allieychen) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: I don't know why we cannot use the file name directly to create the `VariantFile`.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 3]
```diff
@@ -504,6 +508,162 @@ def _get_variant_calls(self, record, formats):
...
+      if k != END_INFO_KEY:
+        if isinstance(v, tuple):
+          info[k] = list(v)
+        else:
+          info[k] = v
+
+    return info
+
+  def _get_variant_calls(self,
+                         samples  # type: libcvcf.VariantRecordSamples
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 1 above)
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 2 above)
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Which file name?

So as discussed offline, `VariantFile` only accepts actual file descriptor, and mocking it is not an option. Therefore, we create pipe to push data in. To use pipe, we need to create child subthread (in the else clause). Ideally, I would supply the entire record list through child thread and write it into the pipe, to get it parsed from VariantFile. However, due to the nature of how `vcf_parser` is structured, the `next()` reads the `text_lines` and parses them line by line, and unless I override it, I can only add logic in the `get_variant()`, where I parse the records on 1-by-1 basis. To achieve this, I create a second pipe:

Now, after supplying all of the header lines into the pipe, child thread waits for new records to come. When `get_variant()` is invoked with a record in the parent thread, it sends the record to child thread via second pipe. Once received, child pushes the record into the first pipe to the VariantFile class in parent thread, from where get_variant reads it back and returns it.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 4]
```diff
@@ -19,10 +19,14 @@
...
+from collections import namedtuple
+
+import sys
+import tempfile
+
 import logging
 import os
-import tempfile
-from collections import namedtuple
+from pysam import libcbcf
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 5]
```diff
@@ -19,10 +19,14 @@
 
 from __future__ import absolute_import
 
+from collections import namedtuple
+
+import sys
+import tempfile
+
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Refactored imports.

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0_bz2_pysam.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
+[
+  {
+    "test_name": "valid-4-0-bz2-pysam",
+    "table_name": "valid_4_0_bz2_pysam",
+    "input_pattern": "gs://gcp-variant-transforms-testfiles/small_tests/valid-4.0.vcf.bz2",
+    "runner": "DirectRunner",
+    "zones": ["us-west1-b"],
+    "vcf_parser": "PYVCF",
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/valid_4_0_bz2_pysam.json]: Wow, that was silly :D 

Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 7]
```diff
@@ -504,6 +504,155 @@ def _get_variant_calls(self, record, formats):
...
+  def _get_variant_calls(self, samples):
+    # type: (libcvcf.VariantRecordSamples) -> List[VariantCall]
+    calls = []
+
+    for (name, sample) in samples.iteritems():
+      if sample.phased:
+        phaseset = DEFAULT_PHASESET_VALUE
+      else:
+        phaseset = None
```
Comment (allieychen) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: simplify to `phaseset = DEFAULT_PHASESET_VALUE if sample.phased else None` if it can fit into one line

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 8]
```diff
@@ -504,6 +504,155 @@ def _get_variant_calls(self, record, formats):
...
+      genotype = None
+      info = {}
+      for (key, value) in sample.iteritems():
+        if key == GENOTYPE_FORMAT_KEY:
+          genotype = list(value) if isinstance(value, tuple) else value
+        elif key == PHASESET_FORMAT_KEY:
+          phaseset = list(value) if isinstance(value, tuple) else value
```
Comment (allieychen) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: The above phaseset value will be override.

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 2 above)
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Hmmm, doing the above refactoring is not difficult, but seems a bit pointless - those methods would have like 1, maybe 2 line each, and would not be repeated anywhere.

Instead, parsed out the child thread into its own method, and added lots of comments. WDYT?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 9]
```diff
@@ -504,6 +506,173 @@ def _get_variant_calls(self, record, formats):
...
+  def _init_with_header(self, header_lines):
+    # The first header line must be similar to '##fileformat=VCFv.*'.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
+
+    # First pipe is repsonsible for supplying data from child thread into the
+    # VariantFile, since it only takes an actual file descriptor as its input.
+    p_read, c_write = os.pipe()
+    # Since vcf_parser is setup to parse 1 line at a time, second pipe is used
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: `# Since vcf_parser processes 1 line at a time, a second pipe...`

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 10]
```diff
@@ -504,6 +506,173 @@ def _get_variant_calls(self, record, formats):
...
+  def _init_child_thread(self, p_read, c_write, c_read, p_write, header_lines):
+    # Child thread's task is to populate data into the pipe that feeds
+    # VariantFile class - first by populating all of the header lines, and then
+    # by redirecting the data received from the second pipe.
+
+    os.close(p_read)
+    os.close(p_write)
+    from_parent = os.fdopen(c_read)
+    to_pysam = os.fdopen(c_write, 'w')
+    # Feed the data with a break line, since we read data line by line - if
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: If we just use `read()`, threads would lock, as it would wait to read until EOF is reached - EOF is not reached, because processes get invoked when next line is 'requested' and wait on each other to get a new record to pass downstream. If the parser wasn't an iterator, we maybe could have just supplied all of the lines to parse, but alas, we cannot do that here.

`..next() invoked`
`......|`
`......V`
`..parent process  ---------------{raw record}---------------->  child process`
`..parent process <--{parsed record}-- VariantFile <--{raw record}--/`
`......|`
`......V`
`..return {parsed record}.`

Given that we cannot rely on EOF to do the reads, we can only rely on EOL, which we utilize via `readline()` instead. Therefore, we make sure that every line to parse has **single** newline char at the end (by trimming and then adding 1 to the end).

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 11]
```diff
@@ -504,6 +506,173 @@ def _get_variant_calls(self, record, formats):
...
+    # we would use the conventional ``read()`` method, both pipes would lock,
+    # while waiting for an EOF signal.
+    for header in header_lines:
+      to_pysam.write(header.strip() + '\n')
+    to_pysam.flush()
+
+    while 1:
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 12]
```diff
@@ -504,6 +504,155 @@ def _get_variant_calls(self, record, formats):
...
+    self._current_line = None
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+    self._to_child = None
+
+  def _init_with_header(self, header_lines):
+    # The first header line must be similar to '##fileformat=VCFv.*'.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 13]
```diff
@@ -504,6 +506,169 @@ def _get_variant_calls(self, record, formats):
...
+    to_pysam.close()
+    sys.exit(0)
+
+  def _init_with_header(self, header_lines):
+    # The first header line must be similar to '##fileformat=VCFv.*'.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
+
+    # First pipe is repsonsible for supplying data from child thread into the
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 9 above)
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

**Comment by tneymanov** [comment]
Ok, now, based on UID I figure out if caller is a super user and weather sudo is required. Works both in local virtual environment with sudo (although will require user to enter their password obviously), and on cloud without it. Fully ready for review.

File: docker/Dockerfile [hunk 14]
```diff
@@ -32,6 +32,18 @@ ADD / /opt/gcp_variant_transforms/src/
 # Needed for installing mmh3 (one of the required packages in setup.py).
 RUN apt install -y g++
 
+# Install Pysam dependencies.
```
Comment (kemp-google) [inline docker/Dockerfile]: I think it's worth calling out that this is only needed because we have a monolithic binary.  Otherwise we only need these dependencies on the Dataflow worker VM.

File: docker/Dockerfile [hunk 15]
```diff
@@ -32,6 +32,18 @@ ADD / /opt/gcp_variant_transforms/src/
...
+RUN apt-get update && apt-get install -y autoconf \
```
Comment (kemp-google) [inline docker/Dockerfile]: put autoconf on the next line and sort please

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 16]
```diff
@@ -504,6 +506,189 @@ def _get_variant_calls(self, record, formats):
...
+  that will feed into PySam object - 1 will feed the data from main process
+  throughout into the child one, while second will get that data in child thread
+  and feed it to PySam library.
+
+  The requirement of using two pipelines comes from the design of VcfParser base
+  class - we could only use a single pipe, but it will divert the parsers.
+  """
+
+  def __init__(self,
+               file_name,  # type: str
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: python style question: we don't line these up we leave them jagged? (the type annotations)

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 17]
```diff
@@ -504,6 +506,189 @@ def _get_variant_calls(self, record, formats):
...
+              ):
+    # type: (...) -> None
+    super(PySamParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    # This member will be properly initiated in _init_with_header().
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Just _vcf_reader?  If so, put the comment after it?

`self._vcf_reader = None  # Initialized in _init_with_header().`

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 18]
```diff
@@ -504,6 +506,189 @@ def _get_variant_calls(self, record, formats):
...
+    self._vcf_reader = None
+    self._to_child = None
+    self._original_info_list = None
+
+  def _init_child_thread(self, p_read, c_write, c_read, p_write, header_lines):
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: These aren't threads they are processes (I thought I'd left that comment but I guess not?)

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 19]
```diff
@@ -504,6 +506,189 @@ def _get_variant_calls(self, record, formats):
...
+    # Child thread's task is to populate data into the pipe that feeds
+    # VariantFile class - first by populating all of the header lines, and then
+    # by redirecting the data received from the second pipe.
+
+    os.close(p_read)
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Why close these streams at the outset?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 20]
```diff
@@ -504,6 +506,189 @@ def _get_variant_calls(self, record, formats):
...
+    p_read, c_write = os.pipe()
+    # Since vcf_parser processes 1 line at a time, a second pipe is used to send
+    # 1 record at a time to the child thread, to pass it downstream back to
+    # parent thread via the first pipe/VariantFile class.
+    c_read, p_write = os.pipe()
+
+    processid = os.fork()
+    if processid:
+      # Parent Thread
+      os.close(c_read)
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Why bother closing these?

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 2 above)
Comment (kemp-google) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: I generally agree with Allie here that it would be nicer if we had

if processid:
  writer()
else:
  reader()

sort of thing.

File: setup.py [hunk 21]
```diff
@@ -40,6 +53,51 @@
...
+
+  def finalize_options(self):
+    pass
+
+  def RunCustomCommand(self, command_list):
+    print 'Running command: %s' % command_list
+    p = subprocess.Popen(
+        command_list,
+        stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    # Can use communicate(input='y\n'.encode()) if the command run requires
```
Comment (kemp-google) [inline setup.py]: Drop these comments

File: setup.py [hunk 22]
```diff
@@ -40,6 +53,51 @@
...
+
+  def initialize_options(self):
+    pass
+
+  def finalize_options(self):
+    pass
+
+  def RunCustomCommand(self, command_list):
+    print 'Running command: %s' % command_list
+    p = subprocess.Popen(
```
Comment (kemp-google) [inline setup.py]: Does python not have a simpler exec form when we don't care about using the standard input?

Comment (tneymanov) [reply setup.py]: Done, but error handling is worse this way.

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 20 above)
Comment (kemp-google) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: It makes it less clear in my mind: you are touching things that you don't need to touch

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 23]
```diff
@@ -504,6 +506,208 @@ def _get_variant_calls(self, record, formats):
...
+    to_pysam.close()
+    sys.exit(0)
+
+  def _init_with_header(self, header_lines):
+    # PySam requires a version header to be present, so add one if absent.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
+
+    # First pipe is responsible for supplying data from child process into the
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: pysam_read, pysam_write ?

variants_read, variants_write ?

I'd explain here that the reason for the variants pipe is that we don't have access to the pysam handle to call into the library from the get_variant function because it's a different process.  It's not really about lines or anything else, it's just that we can't call into pysam from that process.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 24]
```diff
@@ -504,6 +506,187 @@ def _get_variant_calls(self, record, formats):
...
+  def _init_parent_process(self, pysam_read, variants_write):
+    into_pysam = os.fdopen(pysam_read)
+    self._to_child = os.fdopen(variants_write, 'w')
+    self._vcf_reader = libcbcf.VariantFile(into_pysam, 'r')
+    self._original_info_list = self._vcf_reader.header.info.keys()
+
+  def _write_header_lines(self, to_pysam_pipe, header_lines):
+    for header in header_lines:
+      to_pysam_pipe.write(header + '\n')
+    to_pysam_pipe.flush()
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Writelines doesn't add new line characters. Refactored a bit, to make it cleaner.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 25]
```diff
@@ -504,6 +506,187 @@ def _get_variant_calls(self, record, formats):
...
+
+  def _init_child_process(self, variants_read, pysam_write, header_lines):
+    # Child process' task is to populate data into the pipe that feeds
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 26]
```diff
@@ -296,9 +300,56 @@ def __init__(self,
...
+      text_line = next(text_lines).strip()
+      while not text_line.strip():  # skip empty lines.
+        # This natively raises StopIteration if end of file is reached.
+        text_line = next(text_lines).strip()
+      if not self._header_lines[-1].startswith(LAST_HEADER_LINE_PREFIX):
+        raise ValueError('Header missing CHROM line.')
+      header = libcbcf.VariantHeader()
+      for header_line in self._header_lines[:-1]:
+        header.add_line(header_line)
+
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: It would be nice to describe the checks here and why they exist (presumably these are things you found by trial and error than htslib doesn't like?)

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 27]
```diff
@@ -309,13 +360,26 @@ def _process_header_lines(self, header_lines):
       # We need to keep the last line of the header from the file because it
       # contains the sample IDs, which is unique per file.
       header_lines = self._representative_header_lines + header_lines[-1:]
-    self._init_with_header(header_lines)
+    if isinstance(self, PySamParser):
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Same here: a description of why this header munging is needed for the pysam case is helpful

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 28]
```diff
@@ -309,13 +360,26 @@ def _process_header_lines(self, header_lines):
...
+    else:
+      self._init_with_header(header_lines)
 
   def next(self):
-    text_line = next(self._text_lines).strip()
-    while not text_line:  # skip empty lines.
-      # This natively raises StopIteration if end of file is reached.
+    try:
       text_line = next(self._text_lines).strip()
+      while not text_line:  # skip empty lines.
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Can this not be a little function called next_non_empty_line()?  And then we can save a few lines of duplication and drop the comment?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 29]
```diff
@@ -425,8 +489,10 @@ def _convert_to_variant(
       formats: The PyVCF dict storing FORMAT extracted from the VCF header.
         The key is the FORMAT key and the value is
         :class:`~vcf.parser._Format`.
+
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: avoid making unrelated whitespace changes - they just complicate the diff

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 30]
```diff
@@ -504,6 +570,224 @@ def _get_variant_calls(self, record, formats):
...
+  def _parse_to_numeric(self, value):
+    # type: (Any) -> Any
+    if isinstance(value, str) and value.isdigit():
+      return int(value)
+    else:
+      try:
+        return self._parse_float(float(value))
+      except ValueError:
+        # non float
+        if isinstance(value, unicode):
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: What is the unicode case trying to do exactly?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 31]
```diff
@@ -504,6 +570,224 @@ def _get_variant_calls(self, record, formats):
...
+        else:
+          # If a field was not provided in the header, make it a list to
+          # follow the PyVCF standards.
+          info[k] = (
+              self._convert_field(v) if k in self._original_info_list else
+              [self._convert_field(v)])
+
+    return info
+
+  def _parse_to_numeric(self, value):
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: What is this function meant to do?  A comment on the function would be very helpful

It looks like it is trying a variety of conversions to coerce a string (or number?) value into another type of numeric value

Is there no python built in functions for this?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 32]
```diff
@@ -504,6 +570,224 @@ def _get_variant_calls(self, record, formats):
...
+    if precise_value.is_integer():
+      return int(precise_value)
+    return precise_value
+
+  def _convert_field(self, value, numeric=True):
+    # type: (Any) -> Any
+    # PySam currently doesn't recognize '.' value for String fields as missing.
+    if value == '.' or value == b'.' or not value:
+      return None
+    elif isinstance(value, unicode):
```
Comment (kemp-google) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Again I'm not sure what this is doing - what is a 'unicode' object in python and why does it need special handling?

Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Well unicode an 'unencoded' sring. PySam sometimes returns unicode values, sometimes str-s in utf-8 encoding, so I encode all to utf-8, to stay consistent.

Comment (pgrosu) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: @tneymanov But do you know why that is?  There is a very specific reason.

Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Did some analysis over the valid-4.1-large.vcf file - seems like whenever PySam adds a string of its own (like '.' or 'PASS'), or when an INFO field is set to be string, PySam return unicode. In fact the only time it returns str type is for REF alleles...

Is there any other reason? Or any insight why it works like that above?

Comment (pgrosu) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Pretty good first analysis!  To make a short story short, it has to have some consistency among Python versions.  Strings in Python 3 are unicode, while in Python 2 you have to make them unicode.  See: https://portingguide.readthedocs.io/en/latest/strings.html

That's why they mentioned it specifically starting with release 0.8.1 for Python 3 in https://pysam.readthedocs.io/en/latest/release.html#release-0-8-1

That's why you see things like this in [libcutils.pyx](https://github.com/pysam-developers/pysam/blob/master/pysam/libcutils.pyx#L157-L168):

```Python
cdef force_str(object s, encoding=TEXT_ENCODING):
    """Return s converted to str type of current Python
    (bytes in Py2, unicode in Py3)"""
    if s is None:
        return None
    if PY_MAJOR_VERSION < 3:
        return s
    elif PyBytes_Check(s):
        return s.decode(encoding)
    else:
        # assume unicode
        return s
```

Or in [libcbcf.pyx](https://github.com/pysam-developers/pysam/blob/master/pysam/libcbcf.pyx#L151-L154):

```Python
    if PY_MAJOR_VERSION < 3:
        val = s
    else:
        val = PyUnicode_DecodeUTF8(s, strlen(s), NULL)
```

File: setup.py [hunk 33]
```diff
@@ -33,13 +46,52 @@
...
+      self.RunCustomCommand(PYSAM_INSTALLATION_COMMAND)
+
+    except RuntimeError:
+      raise RuntimeError(
+          'PySam installation has failed. Make sure you have the ' + \
+          'following packages installed: autoconf automake gcc libbz2-dev ' + \
+          'liblzma-dev libcurl4-openssl-dev libssl-dev make perl zlib1g-dev')
+
+class build(_build):  # pylint: disable=invalid-name
+  """A build command class that will be invoked during package install.
```
Comment (allieychen) [inline setup.py]: Please leave one blank line between the summary and the additional description.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 34]
```diff
@@ -296,9 +300,68 @@ def __init__(self,
...
+      if not self._header_lines[-1].startswith(LAST_HEADER_LINE_PREFIX):
+        raise ValueError('Header missing CHROM line.')
+      # Construct VariantHeader class from the metadata lines, to verify the
+      # validity of the file. PySam would not raise an error, will omit the
+      # metadata and hide the issue, if the header is incorrect.
+      header = libcbcf.VariantHeader()
+      for header_line in self._header_lines[:-1]:
+        header.add_line(header_line)
+
+      for k, v in header.info.iteritems():
```
Comment (allieychen) [inline gcp_variant_transforms/beam_io/vcf_parser.py]: Does PyVCF raise errors for these cases? I assume PySam does not so we have to check it by ourselves?

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 35]
```diff
@@ -296,9 +300,68 @@ def __init__(self,
...
+      text_line = next(iterator).strip()
+    return text_line
+
+  def _verify_header(self, text_source, range_tracker):
+    try:
+      text_lines = text_source.read_records(self._file_name,
+                                            range_tracker)
+      # Read until first non empty record
+      self._next_non_empty_line(text_lines)
+      if not self._header_lines[-1].startswith(LAST_HEADER_LINE_PREFIX):
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: That's exactly what this section is doing:
It reads the text_line. If line starts with # (ie it's a header), it's skipped so read doesn't happen (due to process_header_lines section). If there are no records, text_lines = text_source... will raise StopIteration right away, which we catch and in that case, if absent, add default CHROM... line (just to "read" an empty file and skip it without complications). However, if text_lines = text_source... read happens, that means that there is at least 1 actual record, and thus we check if CHROM... line exists, and if it does not, raise an error.

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 34 above)
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Yeah, PyVCF does, there are tests for that.

PySam is too smart for its own good, and tries to swallow this issues to allow clients to go through even with incorrect headers. However, for our case, we would want to report header issues upstream to the caller.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 36]
```diff
@@ -504,6 +587,225 @@ def _get_variant_calls(self, record, formats):
...
+      self._to_child.write(data_line.encode('utf-8') + '\n')
+      self._to_child.flush()
+      return self._convert_to_variant(next(self._vcf_reader))
+    except (MemoryError, IOError, ValueError) as e:
+      logging.warning('VCF record read failed in %s for line %s: %s',
+                      self._file_name, data_line, str(e))
+      return MalformedVcfRecord(self._file_name, data_line, str(e))
+
+  def _verify_record(self, record):
+    if record.start < 0:
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Because if pos is for example String for some reason, PySam puts the value as MIN_INT (-2147483648), instead of throwing an error.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 37]
```diff
@@ -504,6 +587,225 @@ def _get_variant_calls(self, record, formats):
...
+          # follow the PyVCF standards.
+          info[k] = (
+              self._convert_field(v) if k in self._original_info_list else
+              [self._convert_field(v)])
+
+    return info
+
+  def _parse_to_numeric(self, value):
+    # Attempt to parse a value to int, then float or return as string.
+    # type: (Any) -> Any
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 38]
```diff
@@ -296,9 +300,68 @@ def __init__(self,
...
+
     self._text_lines = text_source.read_records(self._file_name,
                                                 range_tracker)
 
+  def after_read(self):
+    return
+
+  def _next_non_empty_line(self, iterator):
+    # Get next non-empty stripped record from iterator.
+    # type: (typing.Iterable[str]) -> str
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 36 above)
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcfio.py [hunk 39]
```diff
@@ -261,22 +261,20 @@ def __init__(self,
...
+      self._vcf_parser_class = vcf_parser.PyVcfParser
+    elif vcf_parser_type == VcfParserType.PYSAM:
+      self._vcf_parser_class = vcf_parser.PySamParser
+    else:
+      raise ValueError(
+          'Unrecognized vcf_parser_type: %s.' % str(vcf_parser_type))
 
   def _read_records(self, (file_path, block)):
     # type: (Tuple[str, Block]) -> Iterable(Variant)
     """Reads records from `file_path` in `block`."""
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcfio.py]: Hmmm, sure, but my preferences would had been to have one. Done.

File: gcp_variant_transforms/beam_io/vcfio.py [hunk 40]
```diff
@@ -373,7 +371,7 @@ def __init__(
       desired_bundle_size=DEFAULT_DESIRED_BUNDLE_SIZE,  # type: int
       compression_type=CompressionTypes.AUTO,  # type: str
       allow_malformed_records=False,  # type: bool
-      vcf_parser_type=VcfParserType.PYVCF,
+      vcf_parser_type=VcfParserType.PYVCF, # int
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcfio.py]: Right, my bad. Done.

**Comment by tneymanov** [comment]
Finally ready for review, apologies for the delay.

Changes in the latest commit:
-   Pull latest changes to include region changes which are a necessary now (sorry about that :( ).
-   Modify all instances where PyVCF reader and internal fields are hard-coded to either be ambiguous on which parser is being used, or to rely on specific selection of the parser (either by PySam or PyVCF, since, if Nucleus is selected, PyVCF would had been defaulted).
-   Add integration test for infer headers with PySam ([still running](https://pantheon.corp.google.com/cloud-build/builds/846c7de2-5615-4702-b62f-e19b3f6a393a?project=tural-test-runner)), modify old PySam integration tests to use region instead of zone.
-   Fix a bug where PyVCF and PySam would hang if "#CHROM..." line wasn't provided in with the "merge headers" selection.
-   Modify deployment of integration tests to not install pysam manually - not required, the issue was with other sections of the code.

Note that documentation wasn't properly modified as the PR is enormous as is - docs and comments will be changed in the PR that removes PyVCF and Nucleus entirely.

Currently in the process of modifying the documentation on PySam changes to include additional, header related findings.
Additionally tested DirectRunner infer headers on valid-4.1-large.vcf, now initiating a run on 1k genomes.

**Comment by samanvp** [comment]
Thanks Tural.
It seems this PR presents a transient state and your next PR (which will remove Nucleus and make PySAM as our _only_ VCF parser) will be our long term state. As we discuss offline we better not to spend too much time on this PR and rather focus on the next PR.
